libipt/src/pt_block_decoder.c

```diff
@@ -2515,7 +2522,18 @@ pt_blk_proceed_no_event_fill_cache(struct pt_block_decoder *decoder,
 	 * Cache updates are atomic so even if the two versions were not
 	 * identical, we wouldn't care because they are both correct.
 	 */
-	return pt_bcache_add(bcache, ioff, bce);
+	status = pt_bcache_add(bcache, ioff, bce);
+	if (status < 0)
+		return status;
+
+	/* After we fill a new cache entry, we also need to update the block 
```

Comment:
The comment isn't quite right.  We fill the cache in both cases.  With the inverted check, how about something like:
~~~
/* Filling the cache may have extended @block, so we cannot proceed. */
if (fill_cache)
	return status;

return pt_blk_proceed_no_event_cached(decoder, block, bcache, msec);
~~~